### PR TITLE
generic: add schema for config.yml

### PR DESCRIPTION
**Pull request by ltalirz**
Add JSON schema for validating config.yml

The config.yml is automatically validated against the schema when running `build.sh`. A "modeline" is added to the `config.tpl.yml`, which points to the schema and provides automatic validation in VSCode with the "YAML" extension by RedHat.

### CODE REVIEW

**Comment by ltalirz** [comment]
Hey Xavier, here is a draft for validating the config.yml via a JSON schema
It works directly in VSCode with the YAML extension, just try it on the config.tpl.yml.


Couple of points to decide/fix

- [ ] validation currently fails (`largeviz3d` is actually too long a name)
- [ ] is JSON schema OK or do you prefer to go with a YAML schema (to my understanding, YAML schema is a bit less widely used + not sure whether VSCode will support it)
- [ ] should the schema go in the top level or, e.g. under `/toolset` 
- [ ] the schema currently contains a few fields specific to our setup (e.g. `workstation` VMs). will remove this

File: config.tpl.yml [hunk 1]
```diff
@@ -1,4 +1,8 @@
 ---
+# yaml-language-server: $schema=config.schema.json
+
+# name of the cluster
+project_name: az-hop
```
Comment (ltalirz) [inline config.tpl.yml:5]: Note: this is something we found useful when administering multiple clusters. I can move it out of the PR if you prefer

File: config.schema.json [hunk 2]
```diff
@@ -0,0 +1,929 @@
...
+                },
+                "lustre_base_image": {
+                    "description": "Can be either an image reference or an image_id from the image registry or a custom managed image",
+                    "type": "string"
+                },
+                "lustre_base_plan": {
+                    "description": "The lustre plan to use. Only needed when using the default lustre image from the marketplace. use '::' for an empty plan",
+                    "type": "string"
+                },
+                "workstation_base_image": {
```
Comment (ltalirz) [inline config.schema.json:80]: added by us

File: config.schema.json [hunk 3]
```diff
@@ -0,0 +1,929 @@
...
+                },
+                "scheduler": {
+                    "description": "Scheduler VM configuration",
+                    "$ref": "#/definitions/Cyclecloud"
+                },
+                "cyclecloud": {
+                    "description": "CycleCloud VM configuration",
+                    "$ref": "#/definitions/Cyclecloud"
+                },
+                "workstation": {
```
Comment (ltalirz) [inline config.schema.json:112]: added by us

**Comment by ltalirz** [comment]
Hi @xpillons, thanks for the quick review; I've addressed your comments

File: config.tpl.yml (hunk 1 above)
Comment (xpillons) [reply config.tpl.yml:5]: Yes we have added a cluster_name for slurm to prepare federation. Please work on a PR for a better separation of updates.

**Comment by xpillons** [comment]
@ltalirz please build the schema based strictly on `config.tpl.yml` as it will be confusing otherwise.

**Comment by ltalirz** [comment]
> @ltalirz what would be the process to update the schema ?

For the first draft I inferred the schema from the existing yml file and then manually added descriptions, constraints, etc.

I believe for minor changes to the config.tpl.yml it would be easiest to just edit the schema manually, it's quite straightforward and VS Code already knows all the allowed fields in JSON Schema, which makes it even easier.

One aspect you may want to think about is whether the in-line documentation can eventually be removed from the `config.tpl.yml` since it is present in the schema and is accessible via mouse hover in VS Code:

image

Having it in the schema seems important to be able to provide good error messages; this should also make it straightforward to reuse in a GUI on the Azure Marketplace.

At the moment this information is somewhat duplicated (but maybe ok for the time being).

>  is there a tool to do this ? if so can instructions be added ?

I guess this would call for a new section of the az-hop documentation aimed at developers rather than admins?

**Comment by ltalirz** [comment]
P.S. The schema can also be used to define default values transparently. I've not yet started doing that.

**Comment by xpillons** [comment]
ok, got it. In fact not all customers are using VSCode, hence docs in the template file directly. And yes we should have a developer section in the doc.

**Review by xpillons — APPROVED** [review]
I approve this first release, we can work on adding more constraints in the future

**Pull request merged**